rclpy/rclpy/node.py

```diff
@@ -858,14 +892,22 @@ def _apply_descriptor(
         else:
             descriptor.name = parameter.name
 
-        # The type in the descriptor has to match the type of the parameter.
-        descriptor.type = parameter.type_.value
-
         if check_read_only and descriptor.read_only:
             return SetParametersResult(
                 successful=False,
                 reason='Trying to set a read-only parameter: {}.'.format(parameter.name))
 
+        if descriptor.dynamic_typing or self._allow_undeclared_parameters:
+            descriptor.type = parameter.type_.value
```

Comment:
> Can we allow undeclared parameters, but still have static types? This logic seems to say otherwise.

No, I think I did the same in rclcpp.
The problem is that if there is no parameter declaration, what is the correct type?

IMO the idea of "allow_undeclared_parameters" is that you get a "parameter blackboard" like behavior (i.e. similar to ROS 1), so enforcing that the type doesn't change after the parameter is first set doesn't seem to be a great idea.